enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,182 @@
+
+## Motivation
+
+Upstream is currently moving towards exclusion of in-tree cloud-provider code from `k/k` repository, and instead introducing out-of-tree provider support, making them independent from `k/k` releases, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). Upstream is planning to remove in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).
+
+### Goals
+
+- Prepare openshift components to accomodate the out-of-tree plugins and remove support for in-tree implementation.
+- Add an installer option to select an`external` cloud provider.
+- Deploy cluster infrastructure accordingly to selected `external` option for cloud provider, and allow `cloud-controller` component registration via the provider specific upstream implemetation.
+
+### Non-Goals
+
+- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
+- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.
+- Add support for running multiple cloud configurations in a single cluster.
+
+## Proposal
+
+Add an out-of-tree support based on current IPI, and make a seamless transition for all currently supported providers.
+
+This change will be gradually executed on all suported cloud providers: AWS, Azure, GCP, vSphere and IBM.
+
+### User Stories
+
+As a cloud developer, I’d like to improve support for cloud features for openshift
+
+#### Story 1:
+
+We’d like to deploy control-plane node on a spot instance, to allowing customers to get additional cost reduction on their clusters. In order to introduce and test this feature, we have to implement this in the kubernetes repository, first, then wait for the next release to get a “free” rebase on the feature. We are also bound to the kubernetes release cycles, making it hard to achieve the goal in case the upstream is already going through the feature freeze. Starting downstream is not an option, as the implementation could be rejected upstream or gain less attention then needs to quickly get it into the release, diverging our fork from the source.
+
+#### Story 2:
+
+As an openshift developer I’m assigned a release blocker BZ regarding upstream implementation. Proving the value for merging this and making the bug a release blocker upstream requires careful communication with upstream maintainers, and explaining the value to the people whom are not necessarily involved into internal details, so their opinion could differ and delay merging an important fix.
```

Comment:
I get what you're going for here, but you don't really explain how this proposal is helping in this scenario. I think the value here is we would have better control over the codebase within the team reducing external blockers?

> As an openshift developer responsible for the cloud controllers, I want to have control of the cloud provider code independently from the core of the product so that my team can deliver important fixes in a timely manner without being blocked by the wider community for extended periods.

Reply:
No matter if in- or out-of-tree provider I'd expect a BZ fix land upstream.

Reply:
You can test that in the openshift cluster, but the end goal is we don't diverge more than we need to, and per diverge I mean here we backport upstream accepted fixes to previous releases where we find it required.

Reply:
Added this story. Thanks :)